### DISCUSSION TITLE
5 luglio 2022 - esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#16] [Esami]
Supponiamo di avere un file di 3.200.000 record. Ogni record occupa 250 byte, di cui 40 per il campo chiave.

Ogni blocco contiene 2048 byte. Un puntatore a blocco occupa 5 byte.

Usiamo una organizzazione B-tree con i **blocchi sia del file principale che del file indice pieni al minimo.** 

**Calcolare:**

- il numero di blocchi del file principale?

- il numero di blocchi del file indice?

- il numero di accessi necessari per ricercare un record del file principale?

**Comment by niccolo-fato** [comment]
![IMAGE 2024-01-15 10:41:19](https://github.com/sapienzastudentsnetwork/bdi2324/assets/73995334/d0503fc9-e06a-4723-b817-d7f150c0cd63)

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/a83ae872-8d14-4fb2-a091-0c79bafa7e6b)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/c953c468-d756-423f-bda4-df8b13bc8021)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/7dc4fa67-e769-4990-aba9-8a8a8098c99f)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/bb44e070-583d-4b7a-97f4-7cf2c810e5fe)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/e4466b7a-6470-4046-8cea-6d1579e7efa0)

**Reply by niccolo-fato** [reply]
In RxBLK teoricamente non bisogna prendere la parte intera superiore?

**Reply by Warcophyr** [reply]
si ho notato che ho fatto due errori molto gravi in tutti i B-Tree infatti li sto rifacendo tutti

**Reply by VergariSimone** [reply]
i record (o puntatori) devono entrare per INTERO in un blocco, quindi quando si calcola il
numero di record (o puntatori) per blocco si prende sempre la parte intera INFERIORE se il
numero risulta decimale

**Reply by atyion** [reply]
> i record (o puntatori) devono entrare per INTERO in un blocco, quindi quando si calcola il numero di record (o puntatori) per blocco si prende sempre la parte intera INFERIORE se il numero risulta decimale

Non è sbagliato di per sé quello che hai detto, ma devi controllare ogni volta che il numero rispetti il minimo occupabile, e con la parte inferiore non lo fa mai. Per questo nel caso in cui l'altezza dell'albero è massima (o in generale quando _almeno_ un pezzo di memoria deve essere occupato) si usa la parta superiore direttamente.

**Comment by atyion** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/101365532/a1d5804c-e8e8-416b-af6b-b3572b153222)

**Reply by VergariSimone** [reply]
MBS cosè?

**Reply by atyion** [reply]
> MBS cosè?

Scusate la notazione poco chiara, potevo fare di meglio effettivamente.
E' la blocksize riempita al minimo (50%)

**Reply by VergariSimone** [reply]
ok ma quando devo effettuare questo test? solo per alberi? o anche per hash e isam?

**Reply by atyion** [reply]
> ok ma quando devo effettuare questo test? solo per alberi? o anche per hash e isam?

Secondo me fai prima a evitare il test facendo così: tutti i numeri blocco sono sempre limite superiore, insieme ad altre cose tipo i livelli (per gli alberi), i numeri di accessi (di hash e isam) e i record per bucket (hash). Il resto invece devi vedere se devi raggiungere MASSIMO quel numero o MINIMO quel numero. Normalmente te devi raggiungere AL MASSIMO il 100% della memoria, e quindi usi il limite inferiore, altrimenti se devi ALMENO usare il 50% della memoria (per esempio) devi usare il limite superiore. 
Questa cosa vale per tutte le strutture fisiche da te menzionate.
Potrei essermi dimenticato di menzionare qualche dato che deve avere sempre il limite superiore.
